JAM: Parachain Service Design (#11883)
paritytech/polkadot-sdk: designs/parachain-service-on-jam/parachain-service-on-jam.md

/// `ParaInfo` contributes to the baseline state-balance reservation (see §6.1).
type HeadData = BoundedVec<u8, { 4 * 1024 }>;

/// A validation code reference: its hash plus `pinned` — whether the
/// parachain has *also* solicited it itself, on top of the service's own
/// code-upgrade solicit. See §5.2.
struct ValidationCode {
    hash: ValidationCodeHash,
    pinned: bool,
}

struct ParaInfo {
    /// Current head data (output of last included block).
    head_data: HeadData,
    /// Currently active validation code.
    validation_code: ValidationCode,
    /// Pending code upgrade, if any: the new validation code and the
    /// deadline timeslot after which the upgrade is rejected. See §5.2.
    pending_upgrade: Option<(ValidationCode, Timeslot)>,
    /// Total state balance allocated to this parachain. Set exclusively by
    /// the Coretime chain via `parachain_set_state_balance`. See §6.1.
    total_state_balance: Balance,
    /// State balance currently consumed by this parachain's solicited PVF
    /// preimages (active validation code + pending upgrade, if any).
    /// Increased on `solicit()`, decreased on `forget()`. See §6.1.

>     used_state_balance: Balance,

Reviewer: docs nit: This also includes the cost for staged keys

| Host function | Returns | Purpose |
|---|---|---|
| `export(data: Vec<u8>)` | `u32` | Write a segment to the JAM Data Lake (e.g. outbound XCMP payloads). Returns segment index. |
| `set_parent_head_hash(hash: Hash)` | `()` | Declare the parent head hash this candidate was built on. **Mandatory**: every Refine invocation must call this exactly once or the invocation is invalid (treated as `Err`). The hash is forwarded to Accumulate. |
| `request_code_upgrade(hash: ValidationCodeHash, len: u32)` | `()` | Signal a PVF code upgrade request. See §5.2. |
| `solicit(hash: Hash, len: u32)` | `()` | Mediated forward of JAM's `solicit` (see §6.1). Idempotent — no-op if the parachain is already in `preimage_registry[hash].referencers`. May fail with `InsufficientStateBalance`. For the parachain's own active/pending validation code it only sets `pinned` to true (§5.2). |
| `forget(hash: Hash, len: u32)` | `()` | Mediated forward of JAM's `forget` (see §6.1). Idempotent — no-op if the parachain is not in `preimage_registry[hash].referencers`. May be called for the parachain's own active/pending validation code, where it only sets `pinned` to false (§5.2). |

> | `transfer_out(dest: ServiceId, amount: Balance, memo: Vec<u8>)` | `()` | Transfer balance to another JAM service (Asset Hub only) |

Reviewer: ```suggestion
| `transfer_out(dest: ServiceId, amount: Balance, memo: Memo)` | `()` | Transfer balance to another JAM service (Asset Hub only) |
```

> | `set_authorizer_queue(core: CoreIndex, queue: Vec<AuthorizerHash>, mode: QueueUpdateMode, new_assigner: Option<ServiceId>)` | `()` | Update the authorizer queue for a core (Coretime chain only). `mode` determines whether the queue is applied immediately or cached in service state until the current 80-slot queue is exhausted. `new_assigner`, when `Some`, hands off `assigners[core]` to another service so that service can manage its own core queue going forward; when `None`, the current assigner (Parachain Service) is retained. |

Reviewer: ```suggestion
| `set_authorizer_queue(core: CoreIndex, queue: Vec<AuthorizerHash>, immediate: bool, new_assigner: Option<ServiceId>)` | `()` | Update the authorizer queue for a core (Coretime chain only). `mode` determines whether the queue is applied immediately or cached in service state until the current 80-slot queue is exhausted. `new_assigner`, when `Some`, hands off `assigners[core]` to another service so that service can manage its own core queue going forward; when `None`, the current assigner (Parachain Service) is retained. |
```

This uses mode: `QueueUpdateMode`, but `UpwardMessage::SetAuthorizerQueue` (3.3) and 7.2 use immediate: bool. Alt. define and use `QueueUpdateMode { Immediate, Deferred }`

head_data: BoundedVec<u8, 4096>      = 2 (compact len) + 4096             =   4 098 B
validation_code: ValidationCode = 32 (hash) + 1 (pinned)          =      33 B
pending_upgrade: Option<(ValidationCode, Timeslot)> = 1 + 33 + 4          =      38 B
total_state_balance: Balance                                              =      16 B
used_state_balance: Balance                                               =      16 B
                                                                            -------
                                                                              4 205 B
```

`(ParaId, parachain_log[para_id])` entry, with `parachain_log[para_id]` typed as
`CountedMap<Timeslot, LogEntry, 8>`:

```
LogEntry, Refine variant (dominates Accumulate variant at 992 B):
  enum discriminant                                                       =       1 B
  RefineLogEntry.error: RefineLog::Opaque(BoundedVec<u8, 1024>)
    enum discriminant + 2 (compact len) + 1024                            =   1 027 B
  RefineLogEntry.auth_trace: BoundedVec<u8, 256> = 2 + 256               =     258 B
                                                                            -------

>                                                                               1 286 B

Reviewer: Looks like the largest `AccumulateLog` variant is `StateBalanceUpdateRejected { attempted, current_total, current_used }` = 1 + 3·16 = 49 B, so `AccumulateLogEntry.entries` is 1 (compact len) + 30·49 = 1471 B, plus the `LogEntry` discriminant = 1472 B.

    signature: Signature,
}
```

#### Authorizer Logic

1. Decode config (`pf`) → `authorized_paras`, `collator_set_root`, `collator_set_size`,
   `slot_duration`.
2. Decode token (`pj`) → `collator_proof`, `collator_key`, `signature`.
3. Read the **anchor timeslot** from the refinement context.
4. Compute the expected collator index:
   `collator_index = (anchor_timeslot / slot_duration) mod collator_set_size`.
5. Verify `collator_proof` against `collator_set_root` at leaf `collator_index`,
   confirming `collator_key` is the expected collator for this slot.
6. Verify `signature` over the work package hash using `collator_key`.
7. Return a trace carrying the `collator_key`.

#### Parachain Service Enforcement

Independently of the authorizer code, the Parachain Service's **Refine wrapper** enforces:

- `Vec<ParaId>` (`authorized_paras`) is required to be the first bytes of the config blob.
- `len(authorized_paras) == len(workitems)` — rejects the package if they differ.

#### Anchor Selection and Slot Claiming

The collator picks an anchor block (one of the last 8 JAM blocks) whose timeslot maps
to their collator index. In steady-state AURA the authorizer queue is filled with the

> **same** authorizer hash, so the pool's 8 entries are all the same hash and the collator

Reviewer 2: Not sure about using the same authorizer hash. If the authorizer config included something like `slot % 16`, and we required the anchor slot to be consistent with this, we could prevent a collator from using its assigned slots more than once? It seems like as currently specified this would need to be checked for and punished by accumulate?

Author: Again this is mainly an example for one possible solution. This doesn't mean this has to be THE solution. 

We also discussed what you are proposing and I also thought about it. It is a possible solution. However, then it will require that the Coretime chain holds all the required authorizers and replaces them on JAM over time. (We will probably have more than 80 collators in the future) But yeah, it isn't impossible.

Reviewer 3: I think the guarantors should restrict the anchor to just one JAM block, e.g. the one of the last finished slot. This would resolve a number of issues, e.g. censorship considerations and forcing late (and eventually too late) submission of work packages by core sharing chains. I believe the lenience of the last 8 blocks is for work packages to stay valid through the pipeline, I don't see a reason to allow older anchors at submission time.

Author: What auditors you mean @eskimor?

Reviewer 3: Sorry, meant guarantors.

can pick any of the 8 recent anchors.

> This has a consequence: for **small collator sets** (< 8 collators), a collator could

Reviewer: IIUC the same anchor freedom also lets multiple distinct collators each pick the anchor that elects them and build on the same head (up to 8 when slot_duration==1). The PVF anchor-consistency check catches one collator double-claiming, not N collators each claiming once. So the authorizer can authorize several blocks from different collators on the same parent. Which one is picked as the canonical during accumulation?

Reviewer 4: I think this is worth discussing more, I am also not sure if there is a mechanism proposed to address the freedom of selecting whichever anchor slot within the last 8.

> Which one is picked as the canonical during accumulation?

based on my understanding, whichever one the block author chooses. Validators should only accept one so it's essentially random or can be manipulated probably by a faster collator.

Reviewer 2: Assuming we disallow or punish reuse of a particular anchor slot, and collators use their slot as soon as they can, this situation is only really possible if a collator misses its slot and then tries to use it later (this might happen if it is just lagging)?

Presumably we can disallow/punish use of an old anchor slot to build on a parachain which was last extended using a more recent anchor slot? The problem is probably just that we lose a slot of computation and everything is suddenly lagging by a slot. A collator could possibly fix this itself by using a single slot of on-demand coretime to extend the parachain by two blocks in one slot, that could get quite complicated though. Easier would be to skip a collator slot but there is not really any incentive for a collator to do this? Perhaps rewarding collators more for use of more recent anchor slots could incentivise this enough.

Reviewer 4: > Assuming we disallow or punish reuse of a particular anchor slot, and collators use their slot as soon as they can, this situation is only really possible if a collator misses its slot and then tries to use it later (this might happen if it is just lagging)?

what if there's a malicious collator that always tries to frontrun everybody and submit before the rightful slot author manages to submit (granted that the malicious guy's slot is still within the 8-slot window). Even if the block is invalid, they presumably already wasted the guarantor's time by fetching and validating an invalid work package

Reviewer 2: FWIW the guarantors and in general the JAM validators won't care as they will get paid regardless. But certainly wasting a slot of coretime is not good from the parachain's perspective. If a collator is consistently doing this, ie not using their slot when it becomes available but using it later then really this just shuffles the slots around a bit (someone else will be able to use their slot) and makes things less efficient/more laggy? Doing this as a collator will get you less or no reward, and presumably you will get kicked out eventually. So not sure this is a big deal? I do think we should make the authorizer setup prevent multiple uses of an anchor though, as I commented above, as it reduces the possible extent of this griefing.

Author: I still assume that the guarantor will try to optimize to get stuff it checked on chain. Meaning that work packages with an older anchor will be included before work packages with a newer anchor. Which puts there an ordering. 
If most of the collators are honest, they will produce blocks for which they use the newest JAM block as anchor. So, no malicious author can just put itself before into the queue or whatever. If there is only one honest collator and we use the 24s slot duration, the honest collator will get 4 JAM blocks as possible anchors. In this case we would need to increase the slot duration to 48s. Actually even more to ensure that they can recover the last work package and no one can sneak in another work package.

Same would apply if we have authorizers that only apply to a fixed candidate.

└────────────────────────────────────────────────────────────────────────────┘
```

The CRJA pipeline for a parachain block:

```
[Collect]     Collator gathers transactions, builds a parachain block candidate and puts it into a work package.
    │
    ▼
[Refine]      IN-CORE: Guarantors execute the parachain service refine functionality. Internally this calls the PVF code to verify the work package.
              Stateless, off-chain, metered via PVM gas.
              Output: per-item work-digests plus authorization/export metadata, assembled into a Work Report.
    │
    ▼
[Join]        The Work Report (aggregating per-item work-digests and authorization metadata) is submitted on-chain.
              JAM validators attest (guarantee) its correctness. Availability of the work package is ensured.
    │
    ▼
[Accumulate]  ON-CHAIN: The Parachain Service's Accumulate function runs on-chain.
              It records the new parachain head, applies the PVF's upward host-function
              effects (code upgrades, outbound transfers, authorizer updates, etc.), and queues
              incoming transfers from other services.

```

> ## 3. The Parachain Service

Reviewer: Perhaps is worth adding few lines on the rationale for "one service for all parachains" design choice. I.e. the outcome from your discussion with @zdave-parity . 
E.g. The single-service model gives the host total control over refine: co-scheduling multiple parachains in a single in-core invocation at low latency, SPREE stuff you cited, detecting omitted work items, etc none of which works if each parachain refines in isolation.
With a note: a parachain team that wants to escape that (e.g. to operate outside the polkadot governance) can still launch its own parachain service.

    incoming_transfers: Vec<(ServiceId, Amount, Memo)>,

    /// Per-parachain log. Records both Refine failures (with the
    /// originating authorizer trace) and events recorded during Accumulate
    /// (both errors and positive events). Each parachain keeps at most
    /// 8 entries.
    parachain_log: Map<ParaId, CountedMap<Timeslot, LogEntry, 8>>,

    /// Pending authorizer queue updates that should be applied once the
    /// current 80-slot queue has been consumed.
    pending_authorizer_queues: Map<CoreIndex, Vec<AuthorizerHash>>,

    /// Per-core timeslot at which `assign` was last called for that core.
    /// Combined with the 80-slot queue length, this lets the service compute
    /// when the current queue will be exhausted and a pending queue in
    /// `pending_authorizer_queues` should be applied.
    last_authorizer_assignment: Map<CoreIndex, Timeslot>,

    /// Cross-parachain preimage registry. Holds every preimage the service
    /// has solicited from JAM — including each parachain's active validation
    /// code, any pending-upgrade code, and PVF-initiated `solicit` requests —
    /// under the same referencer-multiplexing scheme. See §6.1.

>     preimage_registry: Map<Hash, PreimageEntry>,

Reviewer: Preimages are requested to JAM via the `(hash, len)` couple. I think we need (Hash, len) as the key here instead of within the PreimageEntry? Same hash with two different lengths are technically provided as two separate things on the JAM layer

    /// preimages (active validation code + pending upgrade, if any).
    /// Increased on `solicit()`, decreased on `forget()`. See §6.1.
    used_state_balance: Balance,
}
```

### 3.2 Work Items

Each work package submitted to the Parachain Service contains one or more **work items**.
For the Parachain Service, a work item represents one parachain candidate. The candidate
itself — validation code hash and PoV — is carried as the work item's **extrinsic data**.

The shape of that extrinsic is:

```rust
struct ParachainCandidate {
    /// The hash of the currently active validation code. Used by Refine to
    /// look up the PVF bytecode from the preimage store.
    validation_code_hash: ValidationCodeHash,

    /// The Proof-of-Validity (PoV) — the actual block data + witness.
    pov: Vec<u8>,
}
```

Initially, each work package will contain a single work item (one parachain candidate).
Support for multiple items per package may be added later.

The `ParaId` for each work item is **not** stored in the work item itself. Instead, it is
sourced from the authorizer config, which is pinned by the Coretime chain (see §7.1). The

> Parachain Service enforces that every authorizer config begins with a `Vec<ParaId>` whose

Reviewer: ```suggestion
Parachain Service enforces that every authorizer config is a `Vec<AuthorizerItem>` whose
```

See proposal at https://github.com/paritytech/polkadot-sdk/pull/11883#discussion_r3460638246

```rust
fn jam_validate_block() -> ()
```

The PVF reads its inputs (PoV, context, downward transfers) through host functions and
writes its outputs (head data, code upgrades, transfers) through host functions. It does
not return a value directly — the `ParachainWorkResult` is assembled by the Parachain
Service's Refine wrapper from the accumulated host-function side effects.

If the PVF exits abnormally (panic, trap, or other failed execution), Refine treats this as
`Err` and records the opaque error payload previously supplied through `report_error(data)` if
one was provided.

The Refine wrapper also fails the invocation as `Err` if the PVF exits without calling
`set_parent_head_hash` exactly once — the parent-head declaration is mandatory.

### 4.3 Host Functions & PVM Imports

On JAM, PVFs execute inside a child PVM instance spawned by the Parachain Service's Refine
function. **Hashing**, **trie operations**, and **signature verification** are expected to
move into PVM guest code — transpilation to native code should bring acceptable performance,
though benchmarks are needed to confirm exact numbers.

> 

Reviewer: To avoid confusion, worth adding a short paragraph stating these are the ABI the Parachain Service defines for the child PVM it spawns to run the PVF, so here the service is the host and the PVF is the guest. Some of these forward 1-to-1 to JAM host functions (the fixed GP set), but most are purely service-defined and have no JAM equivalent.

Author: This document is about the parachain service and not about JAM. I don't really see this confusion.

#### Data access

These forward the full JAM fetch functionality to the PVF:

> | Host function | Returns | Purpose |

Reviewer: ```suggestion
| PVF Host Function | Returns | Purpose |
```

Same in the table below

**Accumulate** entry point of the Parachain Service. This runs on-chain with full access to
service storage.

Accumulate for the Parachain Service covers the parachain-specific parts of what the
relay chain's `enact_candidate` does today; availability, approvals, and disputes are handled
by JAM natively (see §2). The work splits into two categories:

#### Per-work-package work

Performed once for each work package that is being accumulated in this block, in order:

1. **Registration check**: If `para_id` is not in `parachains`, reject the work-package
   immediately and stop processing. In practice the work-report should already have
   failed Refine — the unregistered para's validation code is no longer solicited and
   thus not retrievable via `historical_lookup`, so the result would have come in as
   `Err(InvalidCodeHash)` (see §4.1). But a successful Refine result can still arrive
   at Accumulate for a parachain the Coretime chain cleaned up between guarantee and
   accumulation; this check is the earliest possible reject. No `parachain_log` entry
   is recorded — there is no `parachain_log[para_id]` to append to.

> 2. **Parent head check**: Verify the work result's `parent_head_hash` equals

Reviewer: A collator could, in principle, submit candidates that are valid except with a bad parent to waste guarantor Refine work. This abuse is not free, the collator must hold a valid authorizer slot, so each attempt costs its coretime and gas.
Perhaps we can have an `AccumulateLog` variant for this error kind?

   chunk persists in the buffer until finalization. If this would exceed
   Asset Hub's `total_state_balance`, the append is rejected with
   `AccumulateLog::InsufficientStateBalance` and the buffer is unchanged
   (see §6.1).
2. If `is_last == true`, assembles the full set (prior buffer + `keys`),
   calls JAM `designate` with it, and clears the buffer — releasing the
   prior buffer's footprint. The final chunk goes straight to `designate`
   and never persists in storage, so it needs no headroom check or balance
   charge. If `designate` rejects the call (length not in `valcount`),
   the buffer is still cleared and `AccumulateLog::DesignateRejected` is
   recorded against the Asset Hub `ParaId`. This also gives Asset Hub the
   abort path:
   `set_validator_keys(vec![], true)` flushes a length-zero buffer, which
   `designate` rejects, clearing the staging area.

Asset Hub's `total_state_balance` must include enough headroom for the
worst-case staging footprint (up to `1023 * 336 B ≈ 336 KiB`), provisioned
at genesis. The exact mechanism is TBD.

A worst-case 1023-key rotation takes ~35 Asset Hub work packages (≈ 3.5
minutes at 6 s timeslots).

> 

Reviewer: Worth stating that the staged set must be supplied canonically ordered and deduplicated: order or duplicate keys change the Safrole ring root, so a non-canonical set commits a different (wrong) validator set. 
The JAM's `designate` host function is *dumb* and doesn't reorder or dedup, so this is the caller's responsibility.

  emitted as an `UpwardMessage::SetAuthorizerQueue` from the PVF or from the
  always-accumulate control path.
- JAM's `is_authorized` invocation then checks a work-package token against one of the
  authorizers currently in the core's authorizer pool.

### 7.1 Authorizer Design: AURA Example

The authorizer is a single piece of PVM code (≤ 64 KB) deployed once as a preimage and
reused across all cores. Per-core behavior is controlled by the **config blob** (`pf`),
which is committed to when the authorizer queue is set via `assign`.

#### Config

The config encodes the parachain's collator set and slot timing:

```rust
struct AuthorizerConfig {
    /// Authoritative `ParaId` for each work item in the package, in the same
    /// order as `WorkPackage.workitems`.
    authorized_paras: Vec<ParaId>,
    /// Root of a binary Merkle trie over the collator public keys.
    /// Leaf index == collator index in the set.
    collator_set_root: Hash,
    /// Number of collators in the set.
    collator_set_size: u32,
    /// Slot duration as a multiple of the JAM timeslot (6s).
    /// E.g. slot_duration = 2 means one parachain slot every 12s.

>     slot_duration: u32,

Reviewer: This forces every para in the package to share the same engine config (collator_set_root, collator_set_size, slot_duration). That's fine with one item per package, but breaks once a package carries items for different parachains. 

What about something like:

```rust
/// Parachain consensus data
enum ParaEngine {
    Aura {
        /// Root of a binary Merkle trie over the collator public keys.
        /// Leaf index == collator index in the set.
        collator_set_root: Hash,
        /// Number of collators in the set.
        collator_set_size: u32,
        /// Slot duration as a multiple of the JAM timeslot (x seconds).
        /// E.g. slot_duration = 2 means one parachain slot every 2x.
        slot_duration: u32,
    }
}

/// Per-work-item authorizer entry. `authorized_paras[i]` describes the work
/// item at index `i`: its authoritative `ParaId` plus the engine config the
/// authorizer uses to validate that item's slot claim. With a single work item
/// per package (the current model), `authorized_paras` has length 1.
struct AuthorizerItem {
    /// Authoritative `ParaId` for the work item at this index.
    id: ParaId,
    /// Parachain specific consenus data
    engine: ParaEngine,
}

struct AuthorizerConfig {
    authorized_paras: Vec<AuthorizerItem>
}
````

Of course, this prevents Refine to decode just the list of parachain ids first, without decoding the rest, which I'm not sure if is useful (the consensus engine payload size is **currently** negligible).

Author: This is just an example for a possible authorizer. This is not done to have some generic solution or whatever. Just helping to understand the contexts and how it all works together in the grand picture. So, I don't really see the need for this.

    /// `incoming_transfers`.
    ConsumeTransfersUpTo(u32),
    /// From `parachain_service_upgrade`. See §5.4.
    UpgradeService { code_hash: Hash, min_item_gas: u64, min_memo_gas: u64 },
    /// From `parachain_set_head` — upsert a parachain's head data.
    ParachainSetHead { para_id: ParaId, new_head: HeadData },
    /// From `parachain_set_validation_code` — upsert a parachain's
    /// validation code hash. The service must solicit the validation
    /// code preimage.
    ParachainSetValidationCode {
        para_id: ParaId,
        new_validation_code_hash: ValidationCodeHash,
        new_validation_code_len: u32,
    },
    /// From `parachain_clean_up` — remove all per-parachain state.
    ParachainCleanUp(ParaId),
    /// From `parachain_set_state_balance`. See §6.1.
    ParachainSetStateBalance { para_id: ParaId, new_total: Balance },
}
```

The combined size of all result blobs plus the authorizer trace in a work-report is limited
to **48 KiB** by the Gray Paper.

- **`Ok`** is returned when PVF validation succeeds. The upward host-function calls made
  during Refine (code upgrades, transfers, authorizer updates, etc.) are carried alongside

>   this digest and applied by Accumulate.

Reviewer 5: > are carried alongside this digest 

dq: By this we mean that they are placed in the digest's upward messages? (ie carried as the PVF `Ok` result in `ParachainWorkDigest::Ok::upward_messages`?)

Author: Yes exactly.

| `gas()` | `u64` | Query the remaining gas budget |
| `work_package()` | `WorkPackage` | Access the full encoded work package |
| `work_package_context()` | `RefineContext` | Access the refinement context: anchor (hash, timeslot, posterior state-root), lookup-anchor (hash, timeslot, posterior state-root), prerequisites |
| `auth_config()` | `Vec<u8>` | Access the authorizer config blob |
| `auth_token()` | `Vec<u8>` | Access the authorization token blob |
| `work_items_summary()` | `Vec<WorkItemSummary>` | Summary of all work items (service, code hash, gas limits, counts, payload length) |
| `work_item_summary(index: u32)` | `Option<WorkItemSummary>` | Summary of a specific work item by index |
| `work_item_payload(index: u32)` | `Option<Vec<u8>>` | Payload of a specific work item by index |
| `import_segments()` | `Vec<SegmentMeta>` | Import segments metadata |
| `import_segment(index: u32)` | `Option<Vec<u8>>` | A specific import segment by index |

#### Side-effect host functions

These produce effects carried in the work digest and applied by Accumulate:

| Host function | Returns | Purpose |
|---|---|---|

> | `export(data: Vec<u8>)` | `u32` | Write a segment to the JAM Data Lake (e.g. outbound XCMP payloads). Returns segment index. |

Reviewer 5: dq: Is the `export` function producing effects carried in the work digest in this context? Meaning this is the refine wrapper that calls into JAM's export without touching the digest?

Author: Yes, this is just a pure forward to the native JAM host function.

> | `export(data: Vec<u8>)` | `u32` | Write a segment to the JAM Data Lake (e.g. outbound XCMP payloads). Returns segment index. |

Reviewer 5: For spec-msg specifically we'd need to map an arbitrary `OutgoingMessage { mmr_position: u64, payload: Vec<u8> }` to segments of `SEGMENT_LEN` 4104 bytes. Then, a package might export up to 3072 segments.

This function is simply forwarding to JAM's export (effects applied during Refine), meaning it will write one segment of 4104 bytes?
IIUC, then for variable `OutgoingMessage::payload` we'd need a framing on top of these segments

   staging buffer is reserved worst-case in Asset Hub's baseline footprint
   (§6.1), but an append that would grow the buffer beyond its reserved capacity
   (the `1023`-key bound on `staged_validator_keys`) is rejected as invalid with
   `AccumulateLog::StagedValidatorKeysOverflow`, leaving the buffer unchanged.
2. If `is_last == true`, assembles the full set (prior buffer + `keys`) and
   clears the buffer. The service checks the assembled length against `valcount`:
   if valid it calls JAM `designate` with the set (which goes straight to
   `designate` and never persists in storage); otherwise `designate` is **not**
   called — the length check rejects the set and
   `AccumulateLog::DesignateRejected` is recorded against the Asset Hub `ParaId`.
   This also gives Asset Hub the abort path: `set_validator_keys(vec![], true)`
   yields a length-zero set, which the length check rejects, clearing the staging
   area.

A worst-case 1023-key rotation takes ~35 Asset Hub work packages (≈ 3.5
minutes at 6 s timeslots). State-balance accounting for the staging
buffer is covered in §6.1.

### 5.4 Service Self-Upgrade

> Authority over Parachain Service code upgrades is held by **Asset Hub**. Asset Hub

Reviewer 6: > Authority over Parachain Service code upgrades is held by **Asset Hub**

Just out of curiosity, where will the Parachain Service code be maintained: Fellows, polkadot-sdk, or JAM repo? (I assume, OpenGov+Fellows will have ability to upgrade it?)

How the Parachain Service will be launched/bootstrapped/deployed (service deposit, genesis, trusted AH/Coretime paraIds, privileges, ...)?

---

- AH upgrades Parachain Service
- AH (OpenGov) upgrades Coretime
- Coretime can rescue parachains (incl. AH)

What if Parachain Service is stalled (can this even happen)? 
What if AH (or Coretime) is stalled? 

I created D-Day continuation issue: https://github.com/paritytech/polkadot-sdk/issues/12734 :)

Author: > How the Parachain Service will be launched/bootstrapped/deployed (service deposit, genesis, trusted AH/Coretime paraIds, privileges, ...)?

Coretime/AH are trusted like today, based on their ParaId. Launch is still to be discussed and not part of the spec.

> What if Parachain Service is stalled (can this even happen)?

Everything could stall.  If parachain service stalls, I'm not sure right now on what we could do. I would need to check again the graypaper on how much control an external service can get. This could then do the voting etc.

    Accumulate forwards to JAM upgrade if the preimage is present;
    otherwise it logs AccumulateLog::ServiceUpgradePreimageMissing.

Phase 4: Activate
    On the next JAM invocation the Parachain Service runs under the
    new code.

Phase 5: Forget
    Asset Hub observes the new codehash in Parachain Service state
    and calls forget(old_code_hash, len) (§6.1).
```

---

## 6. Parachain Management

Parachain lifecycle and management is driven by the **Coretime chain**, which owns the
policy layer: ParaId allocation, deposits, and deciding when to create, overwrite, or
clean up a parachain's state.

The Parachain Service exposes four low-level, idempotent host functions that drive
state-balance management, registration, forced updates, and deregistration:

- `parachain_set_state_balance(para_id, new_total)` — set the parachain's quota
- `parachain_set_head(para_id, new_head)` — upsert head data
- `parachain_set_validation_code(para_id, new_validation_code_hash, new_validation_code_len)` — upsert validation code
- `parachain_clean_up(para_id)` — remove all per-parachain state

All four are Coretime-chain-only; the Parachain Service performs no rights-checking

> of its own and in particular **does not enforce ParaId uniqueness** — the Coretime

Reviewer 6: > **does not enforce ParaId uniqueness**

Do we need to think about some XCM changes here? For example, if multiple Parachain Services are running, ParaId is only unique per service instance? How would XCM locations (e.g. bridges) identify a parachain then? (service_id, para_id) instead of Parachain(1000)?

Author: You can already have the same problem today? Same paraids on Kusama and Polkadot? XCM already handles this via the global consensus?

    pov: Vec<u8>,
}
```

Initially, each work package will contain a single work item (one parachain candidate).
Support for multiple items per package may be added later.

The `ParaId` for each work item is **not** stored in the work item itself. Instead, it is
sourced from the authorizer config, which is pinned by the Coretime chain (see §7.1). The
Parachain Service enforces that every authorizer config begins with a `Vec<ParaId>` whose
length matches the number of work items in the package, so that work item `item_index` is
authoritatively bound to `authorized_paras[item_index]`. Refine reads this prefix via
`auth_config()` and uses it to populate `ParachainWorkDigest.para_id`.

### 3.3 Work Digest

The Parachain Service's Refine function returns a parachain work digest per work item.
This digest is forwarded to Accumulate. From the service's perspective, Refine either
succeeds or fails:

```rust
/// The Parachain Service's Refine output for one parachain candidate.
/// Side effects from host functions (code upgrades, transfers, authorizer
/// updates) are recorded separately during Refine and forwarded to Accumulate.

> enum ParachainWorkDigest {

Reviewer 5: dq: Should we version this enum? Coming from the spec-msg perspective we might need initially only a `Provides` hash and later on add a `Requires` which would be a breaking change. There might be other use cases that come up in the future

```rust
enum ParachainWorkDigest {
   V1I(inner_fields)
...
```

Author: `WorkReports` are bound to the service hash. So, they should be any way rejected when the service is upgraded.

absent. Anything a parachain needs to act on reliably belongs either in an `Opaque`
payload its own PVF emitted, or in an accumulate event, which records a state change
that has already happened.

#### Outgoing transfers

Replaying a `TransferOut` (step 7) forwards it to JAM `transfer`. `deferred`
selects between the two modes that host-call offers (Gray Paper, `transfer`):

| | `deferred = None` (plain move) | `deferred = Some((memo, gas))` |
|---|---|---|
| Destination code | none runs | destination's Accumulate runs with `gas` |
| Gas charged | `C_gasT` only | `C_gasT + gas` |
| Balance credited | immediately | when the destination accumulates |
| Requires supervision of `dest` | **yes** | no |

`gas` is charged to the **Parachain Service's own Accumulate gas**, which JAM pools
from the gas limits the block's work items registered for the service. A transfer's
`gas` must therefore be accounted against the limit registered by the candidate that
requested it, so that one parachain cannot spend gas another registered. `transfer_out`
is Asset Hub only, so keeping its demands within that allowance is Asset Hub's responsibility.

> `source` names the debited account, `None` meaning the Parachain Service itself.

Reviewer 7: Assume we leased footprint to a recipient service and want it back. Claw-back via transfer_out (source: Some(recipient), source_supervisor_balance: true) is refused by JAM while it would leave the recipient below its threshold balance, so the part backing its live footprint can't leave until the recipient shrinks its data.

If the service is dead or uncooperative, should be a way to clear its storage?

Reviewer 2: You can clear the storage from the supervisor _but_ only if you know all the storage keys. This requires some kind of archive node which tracks full keys and not just their hashes. PolkaJam does not currently support this...

Author: Sounds like no problem the parachain service is concerned about.

Reviewer 8: I had a chat with @smiasojed and I think it is a valid point.

Asset-hub will have some representation of footprint - so it must have means to force the service to give the footprint back. (no matter what the economy is in the contract - lend, lease, timed allowance).

To enforce this - asset hub needs to remove storage keys. This could be triggered by OpenGov - but asset-hub / parachain-service needs to a mean to do it. Maybe we should add _remove key_ host call to allow this.

In general this probably deserves dedicated ticket to discuss it further.

Author: Yeah you are right. @smiasojed please create an issue in the parachain-service repo. Ty.

Author: Done here: https://github.com/paritytech/polkadot-sdk/pull/11883/changes/15276877084a7aedac57526a7c4fe758ccdc5bef

The Parachain Service is a JAM service whose code implements the on-chain logic of the parachain
host. It holds all per-parachain state and drives the CRJA pipeline for every registered parachain.

The Parachain Service is expected to be an **always-accumulate** service in the Gray Paper sense.
Even in blocks where no parachain candidate becomes available, it still needs an accumulation step
to apply privileged control-plane updates such as authorizer queue changes, validator-key updates,
and other service-level bookkeeping that must take effect without waiting for a parachain block.

This, together with the host calls the service forwards, presupposes three privileged
registrations in JAM's protocol state: membership in the always-accumulate set (with a gas
allowance), being the **delegator** (required for `designate`, §5.3), and being the registered
**assigner** of every core it manages (required for `assign`, §7.1).

### 3.1 Service State Layout

The service state is a key-value store. Logically it contains:

```rust
// Top-level service state (conceptual, not final)
struct ParachainServiceState {
    /// All registered parachains and their current metadata.

>     parachains: Map<ParaId, ParaInfo>,

Reviewer 5: One thing that was of concern for SpecMSG was paraID re-utilization (ie the current para 2222 is deregistered and paraID 2222 is given to new operators)

Similar to polkadot-sdk, we'll assume that Coretime will never reutilize the same ParaID? 

One option to make reutilization work for specmsg is to key by `(ParaID, Reincarnation: u16)`

Author: Not really a concern for the parachain service? Parachain service is not allocating `ParaIds` and `ParaIds` are also not reusable in current Polkadot.

    TransferFailed { id: Compact<u64>, error: TransferError },
    /// A `forget` removed the last referencer without expunging the preimage.
    /// See §6.1.
    ForgetAgainAt { hash: Hash, len: Compact<u32>, due: Timeslot },
    /// `parachain_clean_up` was rejected because the parachain still holds state
    /// beyond its baseline and validation code(s); it must release the rest
    /// first. See §6.4.
    TooMuchStateHeld,
}

/// Why a JAM `transfer` replaying a `TransferOut` failed. See §5.1 step 7.
enum TransferError {
    /// `source` is not a known service.
    UnknownSource,
    /// `dest` is not a known service.
    UnknownDestination,
    /// The service is not `source`'s effective supervisor; only its own regular
    /// balance is exempt. Takes precedence over `DestinationNotSupervised`.
    SourceNotSupervised,
    /// A plain move to another service needs the service to be `dest`'s
    /// effective supervisor. Also covers an identity write (`source == dest`
    /// with both selectors equal).
    DestinationNotSupervised,
    /// The supplied gas is below `dest`'s `min_memo_gas`.
    GasBelowDestinationMinimum,

>     /// The transfer would leave the Parachain Service below its threshold balance.

Reviewer 7: If transfer_out's source is not the Parachain Service, I guess InsufficientServiceBalance applies to the source service, but the doc comment says "the Parachain Service". Should it say source service?

Also: do we get the same error for an account balance that's simply too small and for a balance that would drop below the threshold? Maybe worth splitting them - or carrying the observed figures like StateBalanceUpdateRejected does, so the caller can size a retry.

Author: I updated the doc. Regarding the other question. What should the caller do? If it would fall below `min_balance`, it requires manual intervention to forcefully release stuff as you already said above. If there isn't the expected amount of money in the pot, it means there is a tracking error in the logic having called this. Because nothing else can move funds out of the supervisor pot.

Given all of this, I don't think we should introduce more error messages here.

Phase 2: Verify Solicit
    Asset Hub waits for its next block to be built on top of a JAM
    block whose state reflects the accumulated solicit, then reads
    its parachain_log via the validation inputs and confirms no
    AccumulateLog::InsufficientStateBalance for new_code_hash. If
    insufficient, Asset Hub aborts.

Phase 3: Upgrade
    Asset Hub calls parachain_service_upgrade(new_code_hash, ...).
    Accumulate forwards to JAM upgrade if the preimage is available;
    otherwise it logs AccumulateLog::ServiceUpgradePreimageMissing.

Phase 4: Activate
    On the next JAM invocation the Parachain Service runs under the
    new code.

Phase 5: Forget
    Asset Hub observes the new codehash in Parachain Service state
    and calls forget(asset_hub_para_id, old_code_hash, len) (§6.1).
```

---

### 5.5 Parachain Head Commitment

`accumulate` returns a 32-byte hash. The Parachain Service returns a commitment to
**parachain heads**: each block it builds a binary Merkle tree over the heads that
changed in that block, and returns its root.

```rust
enum MerkleTree {
    Node(Hash, Hash),
    Leaf { para_id: ParaId, head_hash: Hash },
}
```

> - A leaf's `head_hash` is `blake2b-256` of the parachain's `head_data`.

Reviewer 9: Could we use `keccak` here too? saves an on-chain blake2b in the snowbridge verifier.

Author: Yeah :+1:

| `send_upward_message(msg: UpwardMessage)` | `()` | Append one upward message to `ParachainWorkDigest.upward_messages`. Aborts Refine with `Err(RefineLog::UpwardMessagesTooLarge)` if the message would carry the encoded upward messages past the parachain's fixed **40 KiB** budget. Individual variants carry further requirements, documented on the variant. Panics if `msg` fails to decode. |
| `report_error(data: BoundedVec<u8, 1024>)` | `!` | Abort the PVF, failing Refine with `RefineLog::Opaque(data)`. Any bytes beyond 1024 are truncated. Never returns. This is the only way a PVF records a reason for its failure. See §4.2. |

`UpwardMessage` is part of the parachain-visible ABI. Its SCALE encoding is
stable, so a message's `encoded_size()` is computable inside the PVF. The 40 KiB
budget counts the encoded messages alone.

Variants marked **Asset Hub only** or **Coretime chain only** are accepted from
that parachain alone. A variant carrying a `para_id` is further restricted to the
calling parachain, except from the Coretime chain, which may name any parachain
(§6.4). Violating either rule aborts Refine with
`Err(RefineLog::RestrictedHostFunction)`.

> A single Refine invocation may emit at most `MAX_UPWARD_MESSAGES = 1024` upward

Reviewer 5: Should we also bound the total bytes of the digest? In the context of parking/buffering in Accumulate, if we store the full digest, then a single large upward_message might consume the total buffering bytes

  doesn't match any accepted code) or `Opaque(payload)` (the PVF called `report_error`
  before failing, up to 1024 bytes). Accumulate stores the error together with the
  authorizer trace in the per-parachain error log (see §3.1), keyed by the timeslot of the
  JAM block in which Accumulate is running. The entry is deleted when a later successful
  candidate for the same parachain is accumulated **whose lookup-anchor timeslot is
  strictly greater than the stored entry's key**. This can be used for example to slash a
  collator who claimed an authorizer slot that was not theirs.

---

## 4. Refine: In-Core Execution

### 4.1 What Refine Does

Refine is invoked **per work item** by JAM. For each work item at
index `item_index` the Parachain Service performs:

1. Reads the authorizer config via `auth_config()` and decodes the `authorized_paras`
   prefix; enforces `len(authorized_paras) == len(workitems)` and rejects the package
   otherwise (see §7.1).
2. Takes `para_id = authorized_paras[item_index]` as authoritative for this item.
3. Fetches the PVF bytecode via `historical_lookup` (using `validation_code_hash`).

> 4. Instantiates a child PVM with the PVF.

Reviewer 10: Side note: this is out-of-scope of this prototype, but this could be made (probably significantly) more efficient by splitting module compilation and instantiation hostcalls, and merging lookup with compilation into a single hostcall.

cc @zdave-parity

So currently what you'd have to do:

1. fetch the PVF bytecode from the database with `historical_lookup` and write it into the service's address space (at least one copy, possibly two if a fetch from the database need a separate copy)
2. read it back when compiling it into a new PVM module with `machine` (effectively another copy) and write the native machine code into memory (effectively yet another copy, just with on-the-fly translation into native machine code)
3. instantiate the VM with the new module

So that's effectively three copies of the same data back-and-forth at minimum, plus all of the extra processing to recompile it (calculate gas costs from scratch, etc.).

Now, what we could have is something like this:

1. call a `module_handle = compile_from_historical_lookup(id, hash, slot)`
2. the module is fetched, compiled  (same as before), and saved into a PolkaVM-managed persistent cache a handle to this module is returned
3. now it can be instantiated with `instantiate(module_handle)`

But here's the difference: now any subsequent call (per PolkaVM version and unique PVF, since we coul perma-cache this on disk) would look like this:

1. call a `module = compile_from_historical_lookup(id, hash, slot)`
2. the module is already in cache; a handle to already compiled module is returned
3. now it can be instantiated with `instantiate(module_handle)`

This gets rid of the overhead of at least three copies of the data, the recompilation overhead, the gas cost calculation overhead, and the page gas metering costs (copying the data will not only cost time but also extra gas with the new page gas metering), and with good enough engineering could maybe even allow us to assume that `compile_from_historical_lookup` is almost zero-cost (i.e. assume the cache is always hot, since someone would have to constantly upload new code blobs to exploit this, and so we could engineer such attack to be impractical).

------------

Anyway, just my two cents. Once we have this service up and running we could demonstrate the performance difference. (:

Reviewer 2: Hmm I really like this idea, certainly worth making an issue in the GP repo to discuss if you haven't done so yet.

Author: Yeah makes sense :D

though benchmarks are needed to confirm exact numbers.

#### Data access

These forward the full JAM fetch functionality to the PVF:

| Host function | Returns | Purpose |
|---|---|---|
| `lookup(hash: Hash)` | `Option<Vec<u8>>` | Fetch a preimage (e.g. PVF code) |
| `foreign_lookup(service: ServiceId, hash: Hash)` | `Option<Vec<u8>>` | Fetch a preimage from another service's store |
| `gas()` | `u64` | Query the remaining gas budget |
| `work_package()` | `WorkPackage` | Access the full encoded work package |
| `work_package_context()` | `RefineContext` | Access the refinement context (anchor, lookup-anchor, prerequisites) |
| `auth_config()` | `Vec<u8>` | Access the authorizer config blob |
| `auth_token()` | `Vec<u8>` | Access the authorization token blob |
| `work_items_summary()` | `Vec<WorkItemSummary>` | Summary of all work items (service, code hash, gas limits, counts, payload length) |
| `work_item_summary(index: u32)` | `Option<WorkItemSummary>` | Summary of a specific work item by index |
| `work_item_payload(index: u32)` | `Option<Vec<u8>>` | Payload of a specific work item by index |
| `import_segments()` | `Vec<SegmentMeta>` | Import segments metadata |

> | `import_segment(index: u32)` | `Option<Vec<u8>>` | A specific import segment by index |

Reviewer 10: Note: similar to my previous comment, with the current JAM design some of these will also require (unnecessary) extra copies (first copy into the service's memory, then copy into the inner VM memory).

Author: Yeah we would basically need some kind of "forward host call" to child?

    │
    ▼
[Accumulate]  ON-CHAIN: The Parachain Service's Accumulate function runs on-chain.
              It records the new parachain head, applies the PVF's upward host-function
              effects (code upgrades, outbound transfers, authorizer updates, etc.), and queues
              incoming transfers from other services.

```

## 3. The Parachain Service

The Parachain Service is a JAM service whose code implements the on-chain logic of the parachain
host. It holds all per-parachain state and drives the CRJA pipeline for every registered parachain.

The Parachain Service is expected to be an **always-accumulate** service in the Gray Paper sense.
Even in blocks where no parachain candidate becomes available, it still needs an accumulation step
to apply privileged control-plane updates such as authorizer queue changes, validator-key updates,
and other service-level bookkeeping that must take effect without waiting for a parachain block.

### 3.1 Service State Layout

The service state is a key-value store. Logically it contains:

```rust
// Top-level service state (conceptual, not final)

> struct ParachainServiceState {

Reviewer 2: Where are messages between parachains stored?

Author: I skipped messages here completely as written, because we first need a final design. This is kind of orthogonal to this document. My assumption is that we will just shove through some hashes in the future and messages are probably exported to the DA layer.

Reviewer 11: This is the whole point of availability exports in JAM.  They don't save much bandwidth usually, which is why we never did them in Polkadot, but they do save bandwidth here.

    /// Current head data (output of last included block).
    head_data: HeadData,
    /// Hash of the currently active validation code.
    validation_code_hash: ValidationCodeHash,
    /// Pending code upgrade, if any: the new validation code hash and the
    /// deadline timeslot after which the upgrade is rejected. See §5.2.
    pending_upgrade: Option<(ValidationCodeHash, Timeslot)>,
}
```

### 3.2 Work Items

Each work package submitted to the Parachain Service contains one or more **work items**. For the
Parachain Service, a work item corresponds to one parachain candidate:

```rust
/// A work item for the Parachain Service encodes a single parachain candidate.
struct ParachainWorkItem {
    /// The hash of the currently active validation code. Used by Refine to
    /// look up the PVF bytecode from the preimage store.
    validation_code_hash: ValidationCodeHash,

    /// The Proof-of-Validity (PoV) — the actual block data + witness.
    /// This is the large input to Refine. Its upper bound is set by JAM's
    /// work-bundle budget (~13.6 MB, shared with authtoken, authconfig,
    /// extrinsics, and imports) minus whatever the package uses for
    /// everything else.

>     pov: Vec<u8>,

Reviewer 2: Presumably the PoV would be provided as extrinsic data?

Author: Yeah this entire struct should be put into the extrinsics. I will improve the name to make it more clear.

}
```

Initially, each work package will contain a single work item (one parachain candidate).
Support for multiple items per package may be added later.

> The `ParaId` for each work item is **not** stored in the work item itself. Instead, it is

Reviewer 2: This is a bit weird to me. I think ideally authorization should be totally separate from the services that are being executed. In the case of on-demand coretime why would the authorization need a parachain ID?

If you require the auth config to contain a parachain ID for parachain services, then either you need special authorization for parachains vs other things (why?) or you need a parachain ID for non-parachain things which doesn't make sense.

Author: Why do authorizers need to be generic across different services? For me this makes no sense :D 

I have done it this way to bind the authorizer to a ParaId. Point being that otherwise someone could abuse an authorizer to execute something different. For example if collator A is part of the AH collator set. If we do not bind the authorizer to the ParaId, collator A could submit a work package for a different parachain and we could not detect it. (I don't want to match on authorizers in accumulate, which is any way too late and doesn't feel like the right place for this)

Reviewer 2: > Why do authorizers need to be generic across different services?

Authorizers are for cores, not for services. You can have an authorizer that specifically allows you to execute some service, but not sure it makes sense for a service to require use of a particular authorizer. Maybe that's possible but it seems kind of backwards to me.

> Point being that otherwise someone could abuse an authorizer to execute something different

It's fine for an authorizer to only allow execution of particular work-packages. The authorization code has access to the work-package for this reason.

> I don't want to match on authorizers in accumulate, which is any way too late and doesn't feel like the right place for this

I'm saying probably neither refine nor accumulate should care what authorizer was used?

Author: > I'm saying probably neither refine nor accumulate should care what authorizer was used?

Then it should not have access to the authorizer trace. Which is an opaque type and the service can be the only one knowing how to interpret this. 

> Maybe that's possible but it seems kind of backwards to me.

Backwards to what? I mean this doesn't change anything in JAM and there you can still do whatever you want. I'm basically just using what is available.

Author: BTW @zdave-parity if you have better ideas on how to bind a authorizer to a ParaId. Or provide some explanation on why we don't need it. I'm also not fully sold on my design here, but I could not come up with anything better. 

We could also not have the paraids as the output and instead just check that the paraids in the config are matching the paraids stored in the work items, but then we are doubling the space usage (not that wild for 32 byte per paraid, but still wasted space).

Reviewer 2: > We could also not have the paraids as the output and instead just check that the paraids in the config are matching the paraids stored in the work items, but then we are doubling the space usage (not that wild for 32 byte per paraid, but still wasted space).

Yes this is what I was imagining. So refine/accumulate wouldn't care about the authorizer at all. It just seems weird and restrictive to me to require the authorizer to pass on the parachain ID when it may not care about it at all?

> Then it should not have access to the authorizer trace. Which is an opaque type and the service can be the only one knowing how to interpret this.

FWIW I don't know what the intended usage of the authorizer trace is. But it doesn't seem necessary to me here. Maybe I'm misunderstanding something.

Author: > authorizer to pass on the parachain ID when it may not care about it at all?

The authorizer will still care :D Because it would compare authorizer config para_id with the work item para_ids.

Reviewer 2: > The authorizer will still care :D Because it would compare authorizer config para_id with the work item para_ids.

In the case of a static allocation of a core to a parachain yes. But why would the authorizer for some on-demand coretime care about the parachain ID or even that it's being used to execute a parachain vs some other service?

Author: > vs some other service?

I don't get why you are talking about other services? We don't talk about this here and this also doesn't prevent any service to use whatever authorizer they like.

> But why would the authorizer for some on-demand coretime care about the parachain ID

Yeah not really required there, but also doesn't really a problem? It will just work as well.

Reviewer 2: > I don't get why you are talking about other services? We don't talk about this here and this also doesn't prevent any service to use whatever authorizer they like.

Because we presumably want to be able to use on-demand coretime with any service. IDK exactly how on-demand coretime will work but if parachains require an authorizer which outputs a parachain ID then the authorizer for the on-demand coretime will need to do this. I see no reason to require this and it just seems like an annoying constraint from the perspective of the on-demand coretime system?

Author: > Because we presumably want to be able to use on-demand coretime with any service. IDK exactly how on-demand coretime will work but if parachains require an authorizer which outputs a parachain ID

But it does not require this, as I already told you above. You will be able to for example register any kind of authorizer in the coretime chain. This will not enforce anything. Only when the authorizer is being used, it is enforced.

So, you can use whatever authorizer you want for your service.

   authoritative `para_id` (see §4.2).

Because Refine is stateless, it cannot write to service storage.

### 4.2 PVF Entry Point

The Parachain Service's Refine spawns a child PVM and calls the PVF's single entry point:

```rust
fn jam_validate_block() -> ()
```

The PVF reads its inputs (PoV, context, downward transfers) through host functions and
writes its outputs (head data, code upgrades, transfers) through host functions. It does
not return a value directly — the `ParachainWorkResult` is assembled by the Parachain
Service's Refine wrapper from the accumulated host-function side effects.

If the PVF exits abnormally (panic, trap, or other failed execution), Refine treats this as
`Err` and records the opaque error payload previously supplied through `report_error(data)` if
one was provided.

### 4.3 Host Functions & PVM Imports

On JAM, PVFs execute inside a child PVM instance spawned by the Parachain Service's Refine
function. **Hashing**, **trie operations**, and **signature verification** are expected to
move into PVM guest code — transpilation to native code should bring acceptable performance,

> though benchmarks are needed to confirm exact numbers.

Reviewer 11: > transpilation to native code should bring acceptable performance

Oh?  When did this happen?  I'll be surprised if this were the case.  As I understand it, individual chain times seem bound by asymmetric cryptography like signatures, but soon msms and pairings too.

In theory, we could even find that multi-threading msms and pairings gives better performance by liberating the CPU cache faster.  Although tests of that theory could be run on kusma now

Author: This is kind of an assumption. Also JAM doesn't provide any host functions for this right now.

As RISCV supports similar extensions as X86_64, I would assume that we could even support something like vector instructions? 

CC @koute

Reviewer 10: On PVM crypto stuff will definitely be slower than native code currently; by how much exactly - well, we should write a service and benchmark it. (:

Later on we *can* add SIMD instructions to accelerate crypto, but it most likely won't happen for JAM 1.0 (as it would be a pretty big extension to the instruction set, and we do want to *finally* get JAM 1.0 out of the door). That said, even with SIMD inside PVM the crypto implemented as hostcalls will *still* be faster because that can take full advantage of the hardware available to the host (and we can't do this for PVM without making it completely unportable and making it super complex).

So, performance-wise you get: native hostcalls > future PVM with SIMD > current PVM.

Anyhow, it might be worth it to just.... add crypto hostcalls to JAM for whatever important crypto algorithms we use, considering those hostcalls would be relatively simple. But, again, it'd be nice to have some concrete numbers before we make that decision.

Reviewer 11: > Later on we can add SIMD instructions to accelerate crypto

It'd rock for hashing, weird lattice crypto, etc, but..

Do we know if WASM SIMD yields anything resembling the native SIMD?  In PVM & WASM, cryptography would compile using 32 bit limbs, right?  This could change many other optimisations, possibly including nerfing the SIMD.  In popular libraries, we could probably maintain forks that make existing WASM SIMD optimisations available in PVM, but this maybe a higher maintenance burden than hostcalls. 

It's likely curve25519-dalek and/or arkworks have WASM SIMD support now, but if not then some C code somewhere does.

Reviewer 10: > Do we know if WASM SIMD yields anything resembling the native SIMD? In PVM & WASM, cryptography would compile using 32 bit limbs, right? This could change many other optimisations, possibly including nerfing the SIMD. In popular libraries, we could probably maintain forks that make existing WASM SIMD optimisations available in PVM, but this maybe a higher maintenance burden than hostcalls.

WASM SIMD support is irrelevant here, since we're not targeting WASM. (: But in general, its performance *is* significantly lower than pure native.

Anyway, off the top of my head I roughly see three options (note: I haven't explored this in detail yet):

- Use RISC-V vector extensions. Pros: libraries are already starting to support that without us doing anything, Cons: it's not clear what performance we can get, and it's not clear if it can be recompiled well into SIMD on AMD64 (since the model is quite different)
- Add AMD64 SIMD to PVM. Pros: maximum performance, Cons: complex, tricky to support in libraries - we'd have to port them ourselves and/or patch in support in LLVM
- A custom SIMD instruction set (that's somewhere in the middle). This is potentially best of both worlds, assuming we can (by default) recompile RISC-V vector extension into this instruction set, and for critical libraries we could add manual ports.

    │    → when refcount hits 0, forget() is called on the preimage
    │  Drops error_log[para_id], pending_upgrade, and any per-para state
```

Deposits and any economic unwinding are handled by the Coretime chain.

---

## 7. Authorization & Coretime

Coretime on JAM is managed by the **Coretime chain** for **all** services, not just the
Parachain Service. The Coretime chain decides which service — and, for the Parachain
Service, which parachain — owns each core and therefore which authorizer queue should be
installed on it. JAM itself tracks core assignment and coretime usage as protocol state.

For the Parachain Service, the ownership boundary is:

- The **Coretime chain** decides which parachain owns each core and computes the desired
  authorizer queue for that core.
- The **Parachain Service** applies those decisions to JAM via the JAM `assign` host call,
  emitted as an `UpwardMessage::SetAuthorizerQueue` from the PVF or from the
  always-accumulate control path.
- JAM's `is_authorized` invocation then checks a work-package token against one of the
  authorizers currently in the core's authorizer pool.

### 7.1 Authorizer Design: AURA Example

> 

Reviewer 2: It's probably fine to have an AURA-like thing to assign coretime to collators, but I do not think you can rely on this alone. You also need to check collator assignments in accumulate. Authorization concerns whether or not a work-package can be executed on a core or not. It does _not_ determine whether or not a work-package can be accumulated for a service. The latter is presumably important for avoiding censorship etc.

It will (presumably) be possible for anyone to pay for a bit of coretime with their own authorizer, which they can use to execute your parachain and bypass any checks you put in your authorizer (because they're not using it). If you need constraints on who can author blocks and when then these need to be enforced by the parachain accumulation logic directly.

Author: >  You also need to check collator assignments in accumulate

No. For sure the parachain own logic executed in `refine` will check that the collator is actually allowed to build a block. Putting it "into JAM" can actually be done by anyone that has a valid authorizer token. 
Accumulate for sure should not require any special authorizer selection scheme and I don't see any reason it should dictate this one.

Author: BTW, I just talk here about AURA to have some example authorizer scheme to illustrate on how it could work/that it could work at all.

Reviewer 2: > For sure the parachain own logic executed in refine will check that the collator is actually allowed to build a block

Ok that's fine and has the same effect.

> Putting it "into JAM" can actually be done by anyone that has a valid authorizer token. Accumulate for sure should not require any special authorizer selection scheme and I don't see any reason it should dictate this one.

I don't think either refine or accumulate should care what authorizer was used. You are saying that `authorized_paras` will come from the authorizer config? You can't trust the authorizer config so I don't know what the purpose of this is. And it means you can't use any old authorizer, because eg its config needs to have some particular format?

Author: > You can't trust the authorizer config so I don't know what the purpose of this is.

Why can it not trust it? The authorizer hash is `hash(authorizer_code ++ authorizer_config)` which is set when setting the authorizer queue. So, I don't see how it can not trust it

Reviewer 2: If you're checking the authorizer hash in your service code I guess you can. That maybe works for parachains which get a whole core and will only use that core's authorizer. But for parachains using on-demand coretime it surely wouldn't make sense to require a particular authorizer hash?

Author: > If you're checking the authorizer hash in your service code I guess you can

This is not what I'm doing, this is how authorizers work in JAM.

Reviewer 2: > This is not what I'm doing, this is how authorizers work in JAM.

Now I'm confused :sweat_smile:. You can certainly access all the authorizer stuff in your refine function, but if you want to restrict your service to only work with a particular authorizer you have to implement that yourself?

Authorizers are queued for a core, not for a service. A service can be executed on any core. Anyone with some coretime can execute your service whether you like it or not. You can make it so that accumulate doesn't do anything in these cases.

In the parachain case you would want to check that the collator was in the collator set and allowed to build a block in that slot or something like this. You _could_ have the authorizer do these checks and then check in the service logic (refine/accumulate) that a specific authorizer was used. If you need the logic in the authorizer as well then this would save executing the logic twice. Is this what you're saying?

Author: > but if you want to restrict your service to only work with a particular authorizer you have to implement that yourself?

It doesn't do this? It doesn't restrict to one particular authorizer, it only requires that the authorizer returns the `para_id` as first value of the `auth_trace`. I don't get why you think that this is enforcing a particular authorizer.

> In the parachain case you would want to check that the collator was in the collator set and allowed to build a block in that slot or something like this. You _could_ have the authorizer do these checks and then check in the service logic (refine/accumulate) that a specific authorizer was used. If you need the logic in the authorizer as well then this would save executing the logic twice. Is this what you're saying?

I just want that work packages get rejected as early as possible, which is the job of the authorizer ;) Again, I'm not enforcing that any particular authorizer is used, just that the authorizer has one particular return value.

Reviewer 2: > I just want that work packages get rejected as early as possible, which is the job of the authorizer ;)

I disagree. The job of the authorizer is only to check that the core can be used. This _may_ involve checking the work-package, eg if the core is reserved for a specific parachain.

By design _anyone_ can create a work-package for _any_ service and get this accumulated. The service code can make this accumulation not do anything, but it will still happen. The purpose of authorization is only to make sure that this work-package execution has been paid for in some way.

> I don't get why you think that this is enforcing a particular authorizer.

I said that the only way a service can trust the authorizer config is if it enforces use of a particular authorizer. That wasn't the best way of wording it, I meant the only way a service can trust that the authorizer has done something in particular (eg check the collator is in the collator set) is if it enforces uses of a particular authorizer. How can a service trust what the authorizer did otherwise? You said "The authorizer hash is hash(authorizer_code ++ authorizer_config) which is set when setting the authorizer queue", but the authorizer queue is for a core not for a service. There's nothing that requires a particular authorizer queue to be used for a particular service, unless you specifically check for this in your service code and make work-packages which don't use this authorizer queue no-ops in accumulate.

Author: > By design _anyone_ can create a work-package for _any_ service and get this accumulated. The service code can make this accumulation not do anything, but it will still happen. The purpose of authorization is only to make sure that this work-package execution has been paid for in some way.

I don't get why you are mixing here "generic JAM" with the parachain service. For sure, if your service allows anyone to create work packages and they pay for this on their own or you have some authorizers pre-registered, you can do whatever you want. 

However, in the parachain case with AURA we want to prevent for example that someone is wasting a slot they are not assigned to. A slot will be wasted in the moment the authorizer is letting them through. I don't get why you are against preventing this as early as possible. 

Again, this is the design of the parachain service and recommendations for the authorizer. This doesn't change anything with JAM in general.

> That wasn't the best way of wording it, I meant the only way a service can trust that the authorizer has done something in particular (eg check the collator is in the collator set) is if it enforces uses of a particular authorizer.

But even the parachain service itself doesn't care. It just requires that the service is outputting the `para_id` and then it is up to the one paying the core time to put there the correct authorizer. It is also totally fine that I may buy extra core time for asset hub or whatever. Nothing will stop me from doing this.

Reviewer 2: > For sure, if your service allows anyone to create work packages and they pay for this on their own

My point is this is true for _all_ services, by the design of JAM. You can't get away from it. You can only make work-packages you don't want no-ops in the accumulate stage.

> However, in the parachain case with AURA we want to prevent for example that someone is wasting a slot they are not assigned to. A slot will be wasted in the moment the authorizer is letting them through

What do you mean by wasting a slot? If you mean that in the case where you have a core assigned specifically to a parachain, that you don't want that core to ever be used for anything else, then sure. That is what the authorizer is for.

> But even the parachain service itself doesn't care

Then I don't get why you want the parachain ID to be output from the authorizer? Is it just so it doesn't need to be included in the work-package?

Author: > Then I don't get why you want the parachain ID to be output from the authorizer? Is it just so it doesn't need to be included in the work-package?

We are turning in circles :D As already explained, I could add it to the authorizer + the work item. Basically duplicating it. Not that problematic as it is only 32 bytes, but we also already have the auth trace. Again, we need the `para_id` in the `auth_config` to bound the authorizer to the parachain. Preventing that anyone is abusing the authorizer (I already explained all of this above). 

I don't really get why you are focusing that much on this. It doesn't change anything in how JAM is working, it just tries to make parachains work with JAM. If you have a better idea on how to "bound an authorizer" to a parachain, I'm happy to hear.

On JAM, the buffer between Refine and Accumulate is even tighter: the work-report's
combined successful result blobs plus authorizer trace are bounded by **48 KiB**. All
upward messages the PVF emits through host functions have to fit inside that
budget alongside the new head data. Carrying HRMP-style message payloads through the
work-report is therefore not an option — they must go through a different channel, which
is what §8.2 proposes.

### 8.2 Proposed Solution: Full XCMP

The current HRMP model — routing full message payloads through the relay chain — cannot
work on JAM because the work result output is too small to carry message payloads on-chain.
Off-chain messaging is required.

The proposed model is **full XCMP**: only message *headers* and
*hashes* are recorded on-chain; the actual message payloads could be distributed off-chain via
JAM's data availability layer (D3L). This removes the per-message size bottleneck. The
Refine function uses `export()` to write outbound message payloads into DA segments, and
Accumulate only records the message hashes and channel metadata on-chain. See
[paritytech/polkadot-sdk#10449](https://github.com/paritytech/polkadot-sdk/pull/10449)

> for a potential specification of XCMP.

Reviewer 11: So the collators/builders for recieving parachains would fetch the messages from the sending parachain's availability?  Ok  We believe this adds latency, so we could add a faster path, or not.

Author: I would use `export` as some kind of escape hatch, to ensure that the parachain can access the data. For normal operations I would assume that there is a faster way to fetch these messages.

Reviewer 11: There are four possible sources:  sending collators, backers / garentors, availability, and auditors / approval checkers.  And they obtain & offer the block in that order, meaning availability has the block before auditors / approval checkers.

Sending collators could starve the recipient most easily, so they only really make sense if delivery is unreliable, or you trust some fallback.  I've always felt nervous that fallbacks would've buggy code, so I've only suggested these for unreliable delivery.  Backers / garentors are not much better.  Both can send the message really early though.

Availability incurs the cost of contacting 1/3rd of the validators, not really great for a collator, and maybe slow, but always works under our security assumptions.  

Auditors / approval checkers costs only a few connections.  Auditors / approval checkers lack the nice clean threat models of availability, but they do have some reasonable threat model, so they make sense if you want one single easy to debug scheme without fallbacks and without hundreds of network connections.  The problem is they occur last in ELVES, so they might add more latency than availability, even if they run faster and wait on few connections.

I think most parity devs who considered the problems figure they'd do both one of the first two and availability as fallback, and just suffer debugging the code that hardly ever runs in production.

Reviewer: You may want to be explicit here that the DA is the ultimate fallback/correctness path and that a lower-latency delivery path is preferred (e.g. receiver may fetch off-chain from sending collators/guarantors with hash verification)

    ref_count: u32,
}

struct ParaInfo {
    /// Current head data (output of last included block).
    head_data: HeadData,
    /// Hash of the currently active validation code.
    validation_code_hash: ValidationCodeHash,
    /// Pending code upgrade, if any: the new validation code hash and the
    /// deadline timeslot after which the upgrade is rejected. See §5.2.
    pending_upgrade: Option<(ValidationCodeHash, Timeslot)>,
}
```

### 3.2 Work Items

Each work package submitted to the Parachain Service contains one or more **work items**.
For the Parachain Service, a work item represents one parachain candidate. The candidate
itself — validation code hash and PoV — is carried as the work item's **extrinsic data**.

The shape of that extrinsic is:

```rust
struct ParachainCandidate {
    /// The hash of the currently active validation code. Used by Refine to
    /// look up the PVF bytecode from the preimage store.
    validation_code_hash: ValidationCodeHash,

    /// The Proof-of-Validity (PoV) — the actual block data + witness.
    pov: Vec<u8>,
}
```

Initially, each work package will contain a single work item (one parachain candidate).

> Support for multiple items per package may be added later.

Reviewer 12: Why ? I am expecting that for elastic scaling we submit multiple work packages, one for each core. The candidate can indeed have multiple parachain blocks in it (block bundling)

Author: > I am expecting that for elastic scaling we submit multiple work packages, one for each core.

How is this related?

> The candidate can indeed have multiple parachain blocks in it (block bundling)

This is also not related. Block bundling is a parachain "only thing", the relay chain/parachain service doesn't care.

Multiple work items would mean, multiple different parachains sharing the same core.

Reviewer 12: > Multiple work items would mean, multiple different parachains sharing the same core.

I think this usecase needs to be made clear, not just the means to do it. The submitter of the work package in this case is collating for multiple parachains aka `SuperCollator`.  Otherwise we already support core sharing (interlacing), but not at the granularity/efficiency you are proposing here.

    /// - `new_assigner`: when `Some`, hands off `assigners[core]` to the
    ///   given service so it can manage its own queue from that point on;
    ///   when `None`, the current assigner is retained.
    SetAuthorizerQueue {
        core: CoreIndex,
        queue: Vec<AuthorizerHash>,
        immediate: bool,
        new_assigner: Option<ServiceId>,
    },
    /// From `set_validator_keys` — write the upcoming validator keys to
    /// JAM's `stagingset` via `designate`.
    SetValidatorKeys(Vec<ValidatorKey>),
    /// From `consume_transfers_up_to` — prune the consumed prefix of
    /// `incoming_transfers`.
    ConsumeTransfersUpTo(u32),
    /// From `parachain_set_head` — upsert a parachain's head data.
    ParachainSetHead { para_id: ParaId, new_head: HeadData },
    /// From `parachain_set_validation_code` — upsert a parachain's
    /// validation code hash.
    ParachainSetValidationCode { para_id: ParaId, new_validation_code_hash: ValidationCodeHash },
    /// From `parachain_clean_up` — remove all per-parachain state.
    ParachainCleanUp(ParaId),
}
```

The combined size of all result blobs plus the authorizer trace in a work-report is limited

> to **48 KiB** by the Gray Paper.

Reviewer 12: Would it make sense to also include "legacy" support for sending XCM messages? maybe we ca fit a few of these and allow us to launch the service even if offchain XCMP is not yet available.

Author: Depending on how offchain XCMP is progressing we can do this, but for now I would not want to plan with this.

| Host function | Returns | Purpose |
|---|---|---|
| `lookup(hash: Hash)` | `Option<Vec<u8>>` | Fetch a preimage (e.g. PVF code) |
| `foreign_lookup(service: ServiceId, hash: Hash)` | `Option<Vec<u8>>` | Fetch a preimage from another service's store |
| `gas()` | `u64` | Query the remaining gas budget |
| `work_package()` | `WorkPackage` | Access the full encoded work package |
| `work_package_context()` | `RefineContext` | Access the refinement context (anchor, lookup-anchor, prerequisites) |
| `auth_config()` | `Vec<u8>` | Access the authorizer config blob |
| `auth_token()` | `Vec<u8>` | Access the authorization token blob |
| `work_items_summary()` | `Vec<WorkItemSummary>` | Summary of all work items (service, code hash, gas limits, counts, payload length) |
| `work_item_summary(index: u32)` | `Option<WorkItemSummary>` | Summary of a specific work item by index |
| `work_item_payload(index: u32)` | `Option<Vec<u8>>` | Payload of a specific work item by index |
| `import_segments()` | `Vec<SegmentMeta>` | Import segments metadata |
| `import_segment(index: u32)` | `Option<Vec<u8>>` | A specific import segment by index |

#### Side-effect host functions

> These produce effects carried in the work result and applied by Accumulate:

Reviewer 12: It is not clear where these host functions are defined. For example `export` is one defined by JAM, while there rest are not. So it would make sense to add some clarification: **these are not JAM host functions.** , but just "ABI defined by the Parachain Service for the inner PVM"

Author: They are declared exactly here?

Reviewer 12: Yes they are declared here, but the distinction that I pointed out is not explicit. I got confused for a bit while reading.

Reviewer 10: Maybe we should call them something like "service functions" instead of "host functions"? (Since it's not the host that's providing them now, but the service.)

   > `request_code_upgrade`, so the Coretime chain is not on the upgrade path and cannot
   > collect an upgrade-time deposit synchronously. Plausible options include: charging
   > the parachain's own service balance for the second preimage while `pending_upgrade`
   > is set (deducted by Accumulate, requires the service to carry balance); sizing the
   > base registration deposit to cover one concurrent pending upgrade (simple, but makes
   > registration more expensive for parachains that never upgrade); or pre-funding an
   > "upgrade credit" on the Coretime chain that the parachain draws down out-of-band
   > before attempting an upgrade. To be resolved alongside economic modeling of the
   > preimage store.
- **Permissionless submission**: The preimage can be submitted by anyone — the collator,
  block author, or any third party. The JAM protocol validates the hash against the
  solicitation.
- **Timeout protection**: The deadline prevents parachains from indefinitely occupying
  preimage store space with unused code. `UPGRADE_TIMEOUT` is set to **24 hours**, which
  should be sufficient for the preimage to be submitted to JAM after solicitation.

---

> ## 6. Parachain Management

Reviewer 12: When I was reading the Gray Paper I was thinking about this a bit different. So I'd like to propose that here:

### High level
The key take away is that the Parachain Service owns parachain abstraction and the Core Time Chain is a pure marketplace.

The current design places parachain identity and lifecycle inside the Core Time Chain and 4 `Coretime chain only` host functions : `parachain_set_head`, `parachain_set_validation_code`, `parachain_clean_up`, `set_authorizer_queue`.

The Coretime Chain should not know what a parachain is. It should only sell coretime to **services**. The Parachain Service then registers parachains, holds their deposits, and buys coretime from Core Time Chain like any other customer.

This aligns with the graypaper which puts coretime **procurement** on a system parachain, and decouples coretime purchase from the workload that consumes it, so that both pay-as-you-go and reservation models are first-class. The design in this PR currently entangles the two.

### Coretime Chain is a service-agnostic marketplace

Coretime Chain interface now exposes the following: "any JAM service can buy coretime by transferring the appropriate asset with a memo specifying `{ core, authorizer_hash, lease_duration }`." Core Time Chain validates the payment, retains `assigners[core]`, and calls JAM `assign(core, queue)` itself. Core Time Chain never learns `ParaId`, never calls into Parachain Service, never has special host-call privileges in Parachain Service.

The authorizer config is just opaque data from Core Time Chain's perspective. Parachain Service knows the ParaId because Parachain Service authored the config it asked Core Time Chain to install.

### The Parachain Service owns identity,l ifecycle and funds

Parachain Service gains a public registration extrinsic (or upward host-call) gated only by a deposit. Then we need to add some more state like:

- `next_para_id: ParaId`
- `paras: Map<ParaId, ParaInfo>` 
- `accounts: Map<ParaId, Map<AssetId, Balance>>` — parachains can pay with different assets for their blocks
- `supported_assets: BTreeSet<AssetId>` —  what assets we allow parachains to pay with (DOT, PUSD, whatever)
- `price_oracle: ServiceId` —  Not sure how this will work, but we already designing some price oracle chain, should be as service on JAM

The Coretime chain only host calls (`parachain_set_head` / `parachain_set_validation_code` / `parachain_clean_up`) are no longer host calls. These will be just Parachain Service extrinsics (deposit-gated for normal use, governance-gated for emergency recovery). Adn finally `set_authorizer_queue` is no longer a Parachain Service concept at all and only Core Time Chain issues JAM `assign`.

### Multi-token cost model 

Parachain Service exposes an explicit cost model to parachain operators:

struct CostModel {
    /// Reference currency Coretime Chain charges in (likely DOT).
    settlement_asset: AssetId,
    /// Per-block accounting cost charged by Parachain Service to the parachain,
    /// in settlement_asset.
    per_block_cost: Balance,
    /// Allowlist of assets a parachain may fund its account in.
    supported_assets: Vec,
}

Parachain operators top up `accounts[para_id][asset_id]` in any supported asset:

- **Per-block accounting** happens in Parachain Service accumulate when a candidate is successfully enacted: deduct `per_block_cost × oracle_rate(asset → settlement_asset)` from the parachain's balance in their chosen accounting asset. This is just bookeeping , no need to do on-chain swap per block.
- **Coretime chain settlement** happens once per region (much rarer). Parachain Service aggregates demand across parachains, ensures it has enough `settlement_asset` in its own treasury, and issues
`transfer(coretimechain, region_cost, memo=BuyCoretime{...})` for each renewal.
- **Internal swaps** happen out-of-band: when Parachain Service's settlement-asset reserve runs low, it batch-swaps other assets via Asset Hub. 
- **New payment tokens can be added** via Polkadot governance call adding it to `supported_assets`. This means no Parachain Service code change is needed to do this.

Multi-asset is purely a Parachain Service accounting feature/cost model. Other services can basically do the same. Coretime Chain always settles in one currency it understands. This means Parachain Service, not Core Time Chain, is concerned with the FX/oracle risk.

### How the authorizer works

I am not entirely sure this works, but it shoudl be like this:

1. Parachain Service accumulate maintains per-parachain balances.
2. Every region boundary, Parachain service's always-accumulate path iterates parachains and decides who can afford the next region.
3. For parachains that have enough funds: Parachain Service issues `transfer(CoreTimeChain, region_cost, BuyCoretime{core, ah, slots})`. Coretime Chain processes next block, calls JAM `assign`. The AURA authorizer config should not change. The hash entering the queue is the same proposed here.
4. For poor parachains we skips renewal. Coretime Chain's queue for that core is not refilled. Existing pool drains over 8 blocks, then the parachain stops getting core.
5. Per-block costs continue to accumulate in Parachain Service accounting until next region boundary. A parachain that goes briefly negative simply isn't renewed.

Author: > The Coretime Chain should not know what a parachain is. It should only sell coretime to **services**. The Parachain Service then registers parachains, holds their deposits, and buys coretime from Core Time Chain like any other customer.

I'm not sure if you read the document, but for sure the coretime chain is just selling core time to anyone. Parachains/services and whatever. There are still some parachain related things that you need/want to do.

Reviewer 12: Thinking more about it it should not even be a chain, it should be just the `Coretime service`, a system service. As an Web2 analogy IMO it should very much facilitate the access to compute instances (EC2 VM in AWS for example).

Parachain Service and in fact any service requires the compute primitive. The key take away is that the users of these services should not be concerned about core time, similarly to how the users of the DynamoDB service in AWS are not concerned about the VMs the service is running under the hood. 

My proposal allows the Parachain Service to offer free-tier or more flexible pricing models. And generally, I think any service should follow the same pattern and not require their users to buy core-time which is a low level detail, instead they can pay with different assets which offsets the core-time cost allowing the service to also make some profit.

Author: I don't get what you are writing there and how this is related to the parachain service. Yes buying coretime with multiple assets is fine. How the coretime is distributed, if free or whatever is not really relevant. In the end someone needs to pay for it or whatever. Again everything logic that doesn't belong into the service.
